doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,299 @@
+
+The `remap` collection exists in the Materialize time domain, where a `u64`
+indicates milliseconds since the Unix epoch. In the future, if Materialize
+learns to support additional time domains, the `remap` collection will exist in
+the "target" time domain, which may be any one of the available time domains.
+
+Let's take a concrete example. Suppose the `remap` collection for a two
+partition Kafka source contains the following entries:
+
+```
+# (remap,   time,           diff)
+  ((1, +42), 1649686076392,   +1)    # (1)
+  ((2, +40), 1649686076392,   +1)    # (2)
+  ((1, +3),  1649686079487,   +1)    # (3)
+  ((2, +4),  1649686079487,   +1)    # (4)
+```
+
+Entry (1) indicates that offsets 0-42 (inclusive) in partition 1 are assigned a
+Materialize timestamp of Monday, April 11, 2022 14:07:56.392 PM UTC. Entry (2)
+indicates that offsets 0-40 in partition 2 are assigned the same timestamp.
+Entry (3) indicates that offsets 43-45 in partition 1 are assigned a Materialize
+timestamp of 14:07:59.487 UTC on the same day. Entry (4) indicates that offsets
+41-44 in partition 2 are assigned the same timestamp.
+
+Notice how the offsets are described by their *increments*, not their absolute
+value. This allows the collection to be compacted using `partition` as the
+compaction key. This generalizes to the other reclockable source types.
+
+#### `remap` policy
+
+The mapping must always preserve the order of events from the upstream source
+whenever such an order is present. If transactional boundaries are present in
+the upstream source, events in the same transaction must be mapped to the
+same timestamp.
+
+There is explicitly no commitment to the granularity of the mapping. Any number
+of upstream events may be mapped to the same Materialize timestamp, provided the
+above properties are upheld.
+
+Concretely:
+
+  * The mapping for a file source must maintain the line order of the file.
+
+  * The mapping for a Kafka source depends on whether Debezium's transactional
+    metadata topic is in use:
+
+      * If Debezium's transactional metadata topic is not in use, the mapping
+        must maintain the offset order of the Kafka topic within each partition.
+        It does not need to maintain order across partitions.
+
+      * If Debezium's transactional metadata topic is in use, the mapping must
+        maintain the order of transactions as determined by the transactional
+        metadata topic. TODO: can we be more specific here?
+
+  * The mapping for a Kinesis source must maintain order by sequence number
+    within each shard. It does not need to maintain order across shards.
+
+  * The mapping for a PostgreSQL source must maintain order of the commit LSN
+    for each transaction. It must present all changes in a single transaction at
+    the same timestamp. TODO: what is the difference between "end LSN" and
+    "commit LSN" in the PostgreSQL logical replication protocol?
```

Comment:
is it guaranteed that the commit lsn will always appear after all the lsn of all the parts of that transaction? might be a dumb question, but as far as I can tell, lsn is just the order things were written in the binlog

Reply:
Seems fundamental to WALs to me, unless you invent an exotic WAL format. How do you commit a transaction in the WAL if you haven't already written all of its constituent parts to the WAL?

Reply:
I don't want to be too much of a pedant, but .. I think one can have commits before entries, if the commits detail the obligations for the commit. Timely's progress tracking does this, essentially saying "this time is done once you see <these recv acks>". I doubt PG works this way, but it isn't fundamental I think.

Reply:
I didn't know that LSN's represented the state of the postgres WAL, but upon hearing this, it makes sense!

Reply:
> I don't want to be too much of a pedant, but .. I think one can have commits before entries, if the commits detail the obligations for the commit. Timely's progress tracking does this, essentially saying "this time is done once you see ". I doubt PG works this way, but it isn't fundamental I think.

Sorry, yes, that's what I meant by exotic! I've updated my comment to replace "fundamental" with "standard".